db/records/operations/group.py

```diff
@@ -1,123 +1,208 @@
-from sqlalchemy import select, Column, func
+from dataclasses import dataclass, field
+from enum import Enum
+import json
+import logging
+from typing import List, Tuple
+from sqlalchemy import select, func, and_, case, literal
 
-from db.records.exceptions import BadGroupFormat, GroupFieldNotFound
-from db.records.operations.select import get_query, apply_filters
+from db.records import exceptions as rec_exc
 from db.records.utils import create_col_objects
-from db.utils import execute_query
 
+logger = logging.getLogger(__name__)
 
-def append_distinct_tuples_to_filter(distinct_tuples):
-    filters = []
-    for col, value in distinct_tuples:
-        filters.append({
-            "field": col,
-            "op": "==",
-            "value": value,
-        })
-    return filters
+MATHESAR_GROUP_METADATA = '__mathesar_group_metadata'
 
 
-def get_distinct_tuple_values(
-        column_list, engine, table=None, limit=None, offset=None, output_table=None
-):
+class GroupMode(Enum):
+    DISTINCT = 'distinct'
+    PERCENTILE = 'percentile'
+
+
+class GroupMetadataField(Enum):
+    COUNT = 'count'
+    GROUP_ID = 'group_id'
+    FIRST_VALUE = 'first_value'
+    LAST_VALUE = 'last_value'
+
+
+@dataclass(frozen=True, eq=True)
+class GroupBy:
+    column_list: List[str]
```

Comment:
I agree they're a bit clumsy.  One option (that may not have been clear from my writeup) is to avoid the type annotations, and replace them with simple 'docstrings' (We'd need to modify the linter, and some developers' editors may complain).

For now, what if I just re-implement using a normal class with the `@property` getters in the traditional fashion?

Reply:
@kgodey I think we should discuss opportunistic/partial type annotations. I've been really enjoying them. I put some thoughts down about them in an earlier comment: https://github.com/centerofci/mathesar/pull/862#discussion_r764096265